pkg/cortexpb/slicesPool.go

```diff
@@ -0,0 +1,64 @@
+package cortexpb
+
+import (
+	"math"
+	"sync"
+)
+
+const (
+	minPoolSizePower = 5
+)
+
+type byteSlicePools struct {
+	pools []sync.Pool
+}
+
+func newSlicePool(pools int) *byteSlicePools {
+	sp := byteSlicePools{}
+	sp.init(pools)
+	return &sp
+}
+
+func (sp *byteSlicePools) init(pools int) {
+	sp.pools = make([]sync.Pool, pools)
+	for i := 0; i < pools; i++ {
+		size := int(math.Pow(2, float64(i+minPoolSizePower)))
+		sp.pools[i] = sync.Pool{
+			New: func() interface{} {
+				buf := make([]byte, 0, size)
+				return &buf
+			},
+		}
+	}
+}
+
+func (sp *byteSlicePools) getSlice(size int) *[]byte {
+	index := int(math.Ceil(math.Log2(float64(size)))) - minPoolSizePower
+
+	if index >= len(sp.pools) {
+		buf := make([]byte, size)
+		return &buf
+	}
+
+	if index < 0 {
+		index = 0
+	}
+
+	s := sp.pools[index].Get().(*[]byte)
+	*s = (*s)[:size]
+	return s
+}
+
+func (sp *byteSlicePools) reuseSlice(s *[]byte) {
+	index := int(math.Floor(math.Log2(float64(cap(*s))))) - minPoolSizePower
```

Comment:
Should we move this to a helper function?

Reply:
One is Ceil and the other is floor.

On getSlice i wanna get a slice with capacity > than the one i'm allocating (hence ceil).

On the reuseSlice i wanna put the slice on the pool that is AT LEAST the capacity of the slice that i'm reusing:


Ex: getSlice(250) => log2 250 = 7.96 => 2^8 = get from pool with cap = 256

ReuseSlice(250) => log2 250 = 7.96 -> 2^7 => return to the pool with cap(128)

In the case of 256 for instance the pool used will be 256 on both cases.

Reply:
oh ok I see